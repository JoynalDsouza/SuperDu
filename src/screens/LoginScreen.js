import React, {useEffect, useState} from 'react';
import {View, Text, StyleSheet} from 'react-native';
import InputBox from '../components/common/InputBox';
import Button from '../components/common/Button';
import {useQuery, useRealm} from '@realm/react';
import {BSON} from 'realm';
import {rootNavigate} from '../Navigation/navigation';

const LoginScreen = ({navigation}) => {
  const [name, setName] = useState('');
  const realm = useRealm();
  const [error, setError] = useState('');

  const user = useQuery('User') || {};

  const onNameEnter = () => {
    try {
      realm.write(() => {
        realm.create('User', {
          _id: new BSON.ObjectID(),
          name: name,
        });
      });
      onContinuePress();
    } catch (e) {
      console.log(e);
    }
  };

  const resetToDashboard = () => {
    rootNavigate('Home', 'reset');
  };

  const onContinuePress = () => {
    try {
      if (user?.length) {
        resetToDashboard();
      } else {
      }
    } catch (e) {
      console.log(e);
    }
  };

  useEffect(() => {
    onContinuePress();
  }, []);

  return (
    <View style={styles.container}>
      {!user.length ? (
        <View>
          <View
            style={{
              flexDirection: 'row',
              marginHorizontal: 10,
              marginBottom: 16,
            }}>
            <InputBox
              testID={'nameInput'}
              label={'Name'}
              type="alphaNumeric"
              placeholder={'Enter your name'}
              error={error}
              setError={setError}
              inputValue={name}
              setInputValue={setName}
            />
          </View>
          <View style={{alignSelf: 'center'}}>
            <Button
              testID={'enterButton'}
              title={'Enter'}
              onPress={() => {
                onNameEnter();
              }}
              disabled={!!error}></Button>
          </View>
        </View>
      ) : (
        <View>
          <Text>Welcome {user[0]?.name}</Text>
        </View>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {},
});

export default LoginScreen;
