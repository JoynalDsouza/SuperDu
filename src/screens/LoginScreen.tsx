import React, {useEffect, useState} from 'react';
import {View, Text, StyleSheet} from 'react-native';
import InputBox from '../components/common/InputBox';
import Button from '../components/common/Button';
import {useQuery, useRealm} from '@realm/react';
import {BSON} from 'realm';
import {rootNavigate} from '../Navigation/navigation';
import {importRealmData} from '../utils/realm-import-export';
import {User} from '../realm/models/User';
const LoginScreen = ({navigation}) => {
  const [name, setName] = useState('');
  const realm = useRealm();
  const [error, setError] = useState('');

  const user = useQuery(User);

  const onNameEnter = () => {
    try {
      realm.write(() => {
        realm.create('User', {
          _id: new BSON.ObjectID(),
          name: name,
        });
      });
      onContinuePress();
    } catch (e) {}
  };

  const resetToDashboard = () => {
    rootNavigate('Home', 'reset');
  };

  const onContinuePress = () => {
    try {
      if (user?.length) {
        resetToDashboard();
      } else {
      }
    } catch (e) {}
  };

  useEffect(() => {
    onContinuePress();
  }, []);

  return (
    <View style={styles.container}>
      {!user.length ? (
        <View>
          <View
            style={{
              flexDirection: 'row',
              marginHorizontal: 10,
              marginBottom: 16,
            }}>
            <InputBox
              testID={'nameInput'}
              label={'Name'}
              type="alphaNumeric"
              placeholder={'Enter your name'}
              error={error}
              setError={setError}
              inputValue={name}
              setInputValue={setName}
            />
          </View>
          <View style={{alignSelf: 'center', gap: 16}}>
            <Button
              testID={'enterButton'}
              title={'Enter'}
              onPress={() => {
                onNameEnter();
              }}
              disabled={!!error}></Button>

            <Button
              title={'Import App Data'}
              onPress={() => {
                importRealmData(realm);
              }}
            />
          </View>
        </View>
      ) : (
        <View>
          <Text>Welcome {user[0]?.name}</Text>
          <Button title={"Let's Go!!!"} onPress={resetToDashboard}></Button>
        </View>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {},
});

export default LoginScreen;
