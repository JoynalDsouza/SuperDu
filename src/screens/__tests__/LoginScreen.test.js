import React from 'react';
import {render, fireEvent, screen} from '@testing-library/react-native';
import LoginScreen from '../LoginScreen';
import renderer from 'react-test-renderer';

// Mocks
jest.mock('../../Navigation/navigation', () => ({
  rootNavigate: jest.fn(),
}));
jest.mock('@realm/react', () => ({
  useRealm: jest.fn().mockReturnValue({
    write: jest.fn(),
    create: jest.fn(),
  }),
  useQuery: jest.fn(() => {
    return [];
  }),
}));
jest.mock('realm', () => ({
  BSON: {ObjectID: jest.fn().mockReturnValue('mocked-id')},
}));
jest.mock('../../components/common/InputBox', () => 'InputBox');
jest.mock('../../components/common/Button', () => 'Button');

describe('LoginScreen Tests', () => {
  const onContinuePressMock = jest.fn();
  //convert LoginScreen To object

  //   const onContinuePressSpy = jest.spyOn(
  //     LoginScreen.prototype,
  //     'onContinuePress',
  //   );

  //   beforeEach(() => {
  //     // Clear all information stored in the mock spy function before each test
  //     onContinuePressSpy.mockClear();
  //   });
  it('renders correctly when there is no user', () => {
    // Setup
    const {getByText, queryByText, getByPlaceholderText, getByTestId} = render(
      <LoginScreen />,
    );
    const loginScreen = renderer.create(<LoginScreen />).toJSON();
    expect(loginScreen).toMatchSnapshot();
    const element = getByTestId('nameInput');
    expect(element).toBeTruthy();

    expect(queryByText('Welcome')).toBeNull();
  });

  it('updates name state on input change', () => {
    // Setup
    const {getByPlaceholderText, getByTestId} = render(<LoginScreen />);
    const inputBox = getByTestId('nameInput');
    const enterButton = getByTestId('enterButton');
  });

  it('handles "Enter" button press correctly', () => {
    // Setup
    const {getByText, getByTestId} = render(<LoginScreen />);
    const enterButton = getByTestId('enterButton');

    // Act
    fireEvent.press(enterButton);

    // Assertions
    // Check if the realm write function was called, or if `onContinuePress` behaved as expected.
  });

  it('shows welcome message when user exists', () => {
    // Setup
    require('@realm/react').useQuery.mockReturnValue([{name: 'John Doe'}]);
    const {getByText} = render(<LoginScreen />);

    // Assertions
    expect(getByText('Welcome John Doe')).toBeTruthy();
  });

  //   it('calls `onContinuePress` on component mount', () => {
  //     render(<LoginScreen />);

  //     // Assertions
  //     expect(onContinuePressMock).toHaveBeenCalled();
  //   });
});
