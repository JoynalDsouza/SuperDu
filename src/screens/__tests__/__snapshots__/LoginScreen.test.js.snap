// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen Tests renders correctly when there is no user 1`] = `
<View
  style={{}}
>
  <View>
    <View
      style={
        {
          "flexDirection": "row",
          "marginBottom": 16,
          "marginHorizontal": 10,
        }
      }
    >
      <InputBox
        error=""
        inputValue=""
        label="Name"
        placeholder="Enter your name"
        setError={[Function]}
        setInputValue={[Function]}
        testID="nameInput"
        type="alphaNumeric"
      />
    </View>
    <View
      style={
        {
          "alignSelf": "center",
        }
      }
    >
      <Button
        disabled={false}
        onPress={[Function]}
        testID="enterButton"
        title="Enter"
      />
    </View>
  </View>
</View>
`;
